doc/guides/usage-of-primordials-in-core.md

```diff
@@ -0,0 +1,254 @@
+code with primordials should be done with care.
+
+## How to access primordials
+
+The primordials is meant for internal use only, and is only accessible for
+internal core modules. User-code cannot use or rely on primordials, it usually
+fine to rely on ECMAScript built-ins and assume that it will behave as specd.
+
+If you'd like to access the `primordials` object to help you on the Node.js
+development or for tinkering, you can make it globally available:
+
+```bash
+node --expose-internals -r internal/test/binding
+```
+
+## What is in primordials
+
+### Properties of the global object
+
+Object and functions on global objects can be deleted or replaced, using them
+from primordials makes the code more reliable:
+
+```js
+globalThis.Array === primordials.Array; // true
+
+globalThis.Array = function() {
+  return [1, 2, 3];
+};
+globalThis.Array === primordials.Array; // false
+
+primordials.Array(0); // []
+globalThis.Array(0); // [1,2,3]
+```
+
+### Prototype methods
+
+ECMAScript provides a bunch of methods available on builtin objects that are
+used to interact with JavaScript objects.
+
+```js
+const array = [1, 2, 3];
+array.push(4); // Here `push` refers to %Array.prototype.push%.
+console.log(array); // [1,2,3,4]
+
+// %Array.prototype%.push is modified in userland.
+Array.prototype.push = function push(val) {
+  return this.unshift(val);
+};
+
+array.push(5); // Now `push` refers to the modified method.
+console.log(array); // [5,1,2,3,4]
+```
+
+Primordials save a reference to the original method that takes `this` as its
+first argument:
+
+```js
+const {
+  ArrayPrototypePush,
+} = primordials;
+
+const array = [1, 2, 3];
+ArrayPrototypePush(array, 4);
+console.log(array); // [1,2,3,4]
+
+Array.prototype.push = function push(vush) {
+  return this.unshift(val);
+};
+
+ArrayPrototypePush(array, 5);
+console.log(array); // [1,2,3,4,5]
+```
+
+### Safe classes
+
+Safe classes are classes that provide the same API as their equivalent class,
+but whose implementation aims to avoid any reliance on user-mutable code.
+
+### Variadic functions
+
+There are some built-in functions that accepts a variable number of arguments
+(E.G.: `Math.max`, `%Array.prototype.push%`), it is sometimes useful to provide
+the list of arguments as an array. You can use function with the suffix `Apply`
+(E.G.: `MathMaxApply`, `FunctionPrototypePushApply`) to do that.
+
+## Primordials with known performance issues
+
+One of the reason the current Node.js API is not temper-proof today is
+performance: sometimes use of primordials removes some V8 optimisations which,
+when on a hot path, could significantly decrease the performance of Node.js.
+
+* Methods that mutates the internal state of arrays:
+  * `ArrayPrototypePush`
+  * `ArrayPrototypePop`
+  * `ArrayPrototypeShift`
+  * `ArrayPrototypeUnshift`
+* Methods of the function prototype:
+  * `FunctionPrototypeBind`: use with caution
```

Comment:
Should probably add *why* caution should be used.

Reply:
```suggestion
  * `FunctionPrototypeBind`
```